### ISSUE TITLE
Add `bootc apply-live`

### ISSUE THREAD

**Issue by cgwalters** [containers/bootc#76]
This is probably mainly draining the logic from rpm-ostree into ostree-ext, then re-using it here.

But...a whole lot of suddenly OS-specific issues come to the fore.  For example, should we try to distinguish between "new content" and "changes"?

**Comment by cgwalters** [comment]
This one also relates to https://github.com/containers/bootc/issues/165

I know at least some people have also been asking for an "apply live by default" mode.  Note that there's a super tricky detail of that in that if there's kernel changes in the new root, we must in general do something like keep `/usr/lib/modules/$kver` mounted/copied into the running root.

**Comment by alexlarsson** [comment]
> This one also relates to #165
> 
> I know at least some people have also been asking for an "apply live by default" mode. Note that there's a super tricky detail of that in that if there's kernel changes in the new root, we must in general do something like keep `/usr/lib/modules/$kver` mounted/copied into the running root.

Could we bind mount /usr/lib/modules somewhere else like /run/booted-kernel-modules during the initial boot, and then have the kernel also look for modules there? Then we don't have to mess with trying to have union the new and old modules dirs?

**Comment by fabiendupont** [comment]
Could we use the same logic as the [DNF needs-restarting plugin](https://dnf-plugins-core.readthedocs.io/en/latest/needs_restarting.html)? This gives hints whether processes, services or the whole system needs a restart to account for the changes. And this would keep us close to DNF.

It might also be nice to have a `--dry-run` option to know before applying a change whether it will require a process / service / system restart. It would also be interesting to have the ability to use the same logic to compare two images out of band, e.g. in a CI/CD pipeline.

**Comment by SeanMooney** [comment]
> Could we use the same logic as the [DNF needs-restarting plugin](https://dnf-plugins-core.readthedocs.io/en/latest/needs_restarting.html)? This gives hints whether processes, services or the whole system needs a restart to account for the changes. And this would keep us close to DNF.
> 
> It might also be nice to have a `--dry-run` option to know before applying a change whether it will require a process / service / system restart. It would also be interesting to have the ability to use the same logic to compare two images out of band, e.g. in a CI/CD pipeline.

the intent behind apply-live is to avoid reboots, where a kernel update is actually required a more natural fit would be to have a kexec flow to switch directly to the new kernel or defer that until the next actual reboot is requested.

have bootc apply-live ever triver a whole system restart would make it un usabel for some usecases.
i.e. performing software updates when workload cannot be stoped or moved to a different host.
for example, if you have vms or other workloads on the systems in the context fo a cloud computing environment.

A normal user may be okay with downstream, but using this in a data centre context would be very problematic. It would be necessary to live migrate all workloads off a given host to a different one just to update the version of Ping or some other package that is on the host because of a cave.

so the ability to live application without reboot is required for CVE patching and other use-cases.

i hope this issue is prioritized at some point as it's still very important in my view for bootc to be adopted in datacenters.

**Comment by jlebon** [comment]
Just to cross-reference this more explicitly: in OCP we're going to be [implementing something similar to this RFE](https://github.com/openshift/installer/pull/8742) to be able to boot from a disk image which doesn't have e.g. the kubelet and overlay the target node image, which does have it. One success criteria for this RFE should ideally be that we can simplify that logic to use `bootc apply-live` instead. For example, it should ideally work in live systems as well.

**Comment by cgwalters** [comment]
Yeah, last time I tried to update the ostree rust bindings I ran into various issues...it needs some love.

That said, the kexec logic is not *really* complex and we could also reimplement it in Rust directly here.

**Comment by lgarbarini** [comment]
@cgwalters if you're open to a contribution adding separate `kexec` logic directly into bootc I'll see if I can find time in the next couple weeks to give it a try. We'd love to stop waiting for bare metal DDR training/firmware initialization every time I run `bootc upgrade`.

**Comment by cgwalters** [comment]
Yeah, when you are ready to work on it feel free to ping in this issue or on the matrix chat, happy to help

**Comment by Mstrodl** [comment]
> Yeah, last time I tried to update the ostree rust bindings I ran into various issues...it needs some love.
> 
> That said, the kexec logic is not _really_ complex and we could also reimplement it in Rust directly here.

My patch I posted there should fix it enough to let you regenerate the bindings

**Comment by Mstrodl** [comment]
PR merged. I have a tree on my machine with new bindings, I'll see if I can get them out today.

**Comment by Mstrodl** [comment]
Good news! Updated ostree bindings are available (v0.20.0) thanks to @cgwalters :)
https://github.com/ostreedev/ostree/pull/3376
https://github.com/ostreedev/ostree/pull/3378

**Comment by cgwalters** [comment]
Yep and https://github.com/containers/bootc/pull/1069 started using them here

**Comment by jlebon** [comment]
One thing related to this we've been talking about on the MCO side is the idea of lowering [Node Disruption Policies](https://docs.redhat.com/en/documentation/openshift_container_platform/4.17/html/machine_configuration/machine-config-node-disruption_machine-configs-configure) down into RHEL, e.g. in bootc. This would work in tandem with `apply-live` I think, where e.g. `bootc apply-live` would by default only apply live changes if it respects the policies, or error out.

And/or maybe have a `bootc upgrade --apply-or-reboot` which would _either_ apply live changes if all changes respect node disruption policies _or_ reboot the machine.

OTOH I can also see this NDPs integrate with #22 instead, since `/etc` files is probably 95% of what you'd use it for and it's much easier to live apply config changes since those are already distinct from the host image and actually overlaid live.

But on the other _other_ hand, having the concept of NDPs be generic enough so it applies to both OS content and config overlays seems like it'd be useful.

**Comment by evan-goode** [comment]
What's the status of this? It would be awesome to have for https://gitlab.com/fedora/bootc/tracker/-/issues/4.

**Comment by cgwalters** [comment]
I think it should be very straightforward to just overmount a new `/` composefs - at least in the case where it's read only. The tricker questions are around `/etc` because if we just overmount that then anything with an open fd will be silently writing to an old mount, and it'd be pretty easy to get data loss. A lot of the problems here are similar to [mutable sysexts](https://www.freedesktop.org/software/systemd/man/latest/systemd-sysext.html#Mutability); hard to avoid giving people a choice of bugs. The rpm-ostree logic here does an eager 3 way merge which is *mostly* what people want I think, but does mean that "apply live" also creates a persistent mutation.

For use cases that can have `/etc` readonly (not just transient, very viable with more stateless systems) then it's all much simpler because in that case `/etc` isn't even a mount point, it's just part of `/` too.

**Comment by MatthiasGrandl** [comment]
Any rough idea when this is going to land? Once it does I will convert all my machines to using bootc hehe. Mind you soft-reboot is already great, but no unnecessary service restarts would make it much more viable for my usecase :)

**Comment by Mstrodl** [comment]
Would love to see this in ostree, then called out to from bootc if it's practical

**Comment by Jannik2099** [comment]
We are looking into bootc for immutable base images for our clusters and would *really* benefit from this feature - rebooting for every update is simply not an option.

For now, I'm trying to hack something together with systemd-sysext, as bootc seems to be the ideal solution for us otherwise.

**Comment by jmarrero** [comment]
Another use case that came up for this, is logical bound images.

**Comment by cgwalters** [comment]
> Another use case that came up for this, is logical bound images.

Yeah. While it is not ergonomic right now, if one wants to run a newer/different version of a LBI, you can do this by having a dynamic configuration which temporarily masks the builtin quadlet and overrides it with e.g. one in `/run`.
